### PR TITLE
parallel networks for SAC

### PR DESCRIPTION
**Pull request by hnyu**

### CODE REVIEW

**Comment by emailweixu** [comment]
Currently, QNetwork does not implement make_parallel. Thus for discrete actions, use_parallel_network does not have any effect.

**Comment by emailweixu** [comment]
> Currently, QNetwork does not implement make_parallel. Thus for discrete actions, use_parallel_network does not have any effect.

Can be dismissed. You actually implemented ParallelQNetwork

File: alf/layers.py [hunk 1]
```diff
@@ -126,6 +126,11 @@ def __init__(self,
             bias_init_value (float): a constant
         """
         super(FC, self).__init__()
+        # get the argument list with vals
+        self._kwargs = locals()
```
Comment (emailweixu) [inline alf/layers.py]: suggest to make a copy of locals(). Otherwise, future addition of local variable will change self._kwargs

Comment (hnyu) [reply alf/layers.py]: updated

File: alf/networks/critic_networks.py [hunk 2]
```diff
@@ -171,19 +174,18 @@ class ParallelCriticNetwork(Network):
...
-                 name='ParallelCriticNetwork'):
+                 name="ParallelCriticNetwork"):
         """
         It create a parallelized version of ``critic_network``.
 
         Args:
             critic_network (CriticNetwork): non-parallelized critic network
             n (int): make ``n`` replicas from ``critic_network`` with different
                 initialization.
-            name (str): name of this network
```
Comment (Haichao-Zhang) [inline alf/networks/critic_networks.py:182]: why remove this comment?

File: alf/networks/q_networks.py [hunk 3]
```diff
@@ -20,14 +20,14 @@
...
-from alf.networks import EncodingNetwork, LSTMEncodingNetwork
-from alf.networks import PreprocessorNetwork
+from alf.networks import EncodingNetwork, LSTMEncodingNetwork, ParallelEncodingNetwork
+from alf.networks import Network
 from alf.tensor_specs import TensorSpec, BoundedTensorSpec
 import alf.utils.math_ops as math_ops
 
 
 @gin.configurable
-class QNetwork(PreprocessorNetwork):
```
Comment (Haichao-Zhang) [inline alf/networks/q_networks.py:30]: We need to derive from ```PreprocessorNetwork``` instead of ```Network```?

**Review by emailweixu — APPROVED** [review]

File: alf/networks/critic_networks.py (hunk 2 above)
Comment (hnyu) [reply alf/networks/critic_networks.py:182]: It's quite obvious

File: alf/networks/q_networks.py (hunk 3 above)
Comment (hnyu) [reply alf/networks/q_networks.py:30]: preprocessing is handled by `self._encoding_net`; so no need to derive from that

**Pull request merged**